From: Michal Hocko <mhocko@suse.cz>
Subject: pm, hinernate: use put_page in release_swap_writer
Patch-mainline: no, suse specific
References: bnc#943959

Martin Wilck from Fujitsu has reported the following bad page state issue:
[  308.170359] PM: Image saving progress:  90%
[  311.215595] PM: Image saving progress: 100%
[  311.296947] PM: Image saving done.
[  311.301502] PM: Wrote 12867096 kbytes in 33.13 seconds (388.38 MB/s)
[  311.310674] PM: S|
[  311.321074] BUG: Bad page state in process bash  pfn:3fe0d85
[  311.327917] page:ffffea00df92f518 count:0 mapcount:0 mapping:          (null) index:0x0
[  311.337314] page flags: 0x6fffff80004008(uptodate|waiters)
[  311.344005] page dumped because: PAGE_FLAGS_CHECK_AT_FREE flag(s) set
[  311.351744] bad because of flags:
[  311.355977] page flags: 0x4000(waiters)
[...]
[  311.441820] Supported: Yes
[  311.445547] CPU: 27 PID: 11846 Comm: bash Tainted: G           OE X 3.12.28-4-default #1
[  311.455315] Hardware name: FUJITSU PRIMERGY RX4770 M2/D3349-A1, BIOS V5.0.0.9 R1.7.0 for D3349-A1x                      06/29/2015
[  311.469919]  ffffffff817f9a98 ffffffff8150b1db ffffea00df92f518 ffffffff81506c55
[  311.478903]  ffffea00df92f518 0000000000000000 000fffff80000000 ffffffff8113d265
[  311.487873]  fff000007c000000 ffffea00df92f518 006fffff80004008 0000000000000000
[  311.496865] Call Trace:
[  311.500265]  [<ffffffff8100467d>] dump_trace+0x7d/0x2d0
[  311.506883]  [<ffffffff81004964>] show_stack_log_lvl+0x94/0x170
[  311.514276]  [<ffffffff81005d91>] show_stack+0x21/0x50
[  311.520798]  [<ffffffff8150b1db>] dump_stack+0x41/0x51
[  311.527305]  [<ffffffff81506c55>] bad_page.part.66+0xe8/0x102
[  311.534499]  [<ffffffff8113d265>] free_pages_prepare+0x165/0x180
[  311.541979]  [<ffffffff8113dc4c>] free_hot_cold_page+0x2c/0x180
[  311.549363]  [<ffffffff810a62cf>] swsusp_write+0x17f/0x3e0
[  311.556251]  [<ffffffff810a1241>] hibernate+0x181/0x1d0
[  311.562849]  [<ffffffff8109efac>] state_store+0xcc/0xd0
[  311.569442]  [<ffffffff8120bc5e>] sysfs_write_file+0xbe/0x140
[  311.576608]  [<ffffffff8119e9f8>] vfs_write+0xb8/0x1e0
[  311.583087]  [<ffffffff8119f418>] SyS_write+0x48/0xa0
[  311.589459]  [<ffffffff81519329>] system_call_fastpath+0x16/0x1b
[  311.596905]  [<00007f97875594e0>] 0x7f97875594df

The PG_waiters bit is stale and it should have been cleared before the page was
freed. This is normally done in put_page path so let's use put_page rather
thatn free_page in release_swap_writer directly.

Signed-off-by: Michal Hocko <mhocko@suse.cz>

---
 kernel/power/swap.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/kernel/power/swap.c
+++ b/kernel/power/swap.c
@@ -310,7 +310,7 @@ static int write_page(void *buf, sector_
 static void release_swap_writer(struct swap_map_handle *handle)
 {
 	if (handle->cur)
-		free_page((unsigned long)handle->cur);
+		put_page(virt_to_page((unsigned long)handle->cur));
 	handle->cur = NULL;
 }
 
